### ISSUE TITLE
Cosine Similarity of Anchor and Negative is not taken into consideration in Loss calculation

### ISSUE THREAD

**Issue by ashokrajab** [xlang-ai/instructor-embedding#94]
In train.py def compute_loss() the cosine similarity value of query vs positive is first appended to all_scores. Following that cosine similarity of query vs negative is appended to `all_scores`. This is done for every triplet in the batch. Finally, the `all_scores` shape will be (batch_size x batch_size+1). 
Then when calculating the loss = nn.CrossEntropyLoss()(all_scores, labels) the labels are entirely zeros.

This means only the cosine similarity of index 0 in all_scores will be taken into consideration of loss computation (in other words loss = - log(all_scores[0])). The cosine similarity between query and negatives does not contribute to loss computation.

In my opinion, the cosine similarity value should have been passed through a sigmoid function and then Binary Cross Entropy Loss function should have been used (based on whether the similarty is for query vs pos or query vs neg).

@Harry-hash @hongjin-su Could you please shed some light on this?

**Comment by hongjin-su** [comment]
Hi Thanks a lot for your interest in the INSTRUCTOR!

I think nn.CrossEntropyLoss has integrated the sigmoid function and considered cosine similarity between query and negatives, according to the documentation in: https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html.

To further verify, you may try the following script and see that `loss` and `loss1` are always the same:
```
import torch

l = torch.nn.CrossEntropyLoss()
logits = torch.randn(4,5)
labels = torch.zeros(logits.size(0)).long()

loss = torch.nn.CrossEntropyLoss()(logits, labels)
print(loss)

loss1 = 0
for logit in logits:
    loss1 += torch.log(torch.exp(logit[0])/torch.sum(torch.exp(logit)))
loss1 /= logits.size(0)
loss1 = -loss1
print(loss1)
```

**Comment by ashokrajab** [comment]
Thank you for the response @hongjin-su.
Okay.. I now understand that I misunderstood the construction of `all_scores` in `compute_loss`. I now understand that it will be a nx1 tensor. Therefore, using label value as 0, all the similarity scores will be taken into CrossEntropy loss calculation.

But I still have concern with the loss calculation. When we feed the cosine similarity calculated for query vs negative to the CrossEntropy loss function, it will try to maximize the similarity value. But we do not want that. We want to minimize the similarity for query vs negative. Thats why I was thinking whether BinaryCrossEntropy should have been used with label as 1 for query vs positive and label as 0 for query vs negative. 
This way we maximize the similarity cross for query vs positive and minimize the similarity score for query vs negative.

Please correct me if I'm wrong.
